### ISSUE TITLE
add support for maybe structs and maybe lists

### ISSUE THREAD

**Issue by gcanti** [gcanti/tcomb-form#236]
Let's say you have this type:

```js
const Type = t.struct({
  name: t.String,
  address: t.struct({
    street: t.String,
    city: t.String
  })
});
```

everything's ok, all fields are required. Now consider this type:

```js
const Type = t.struct({
  name: t.String,
  address: t.maybe(t.struct({
    street: t.String,
    city: t.String
  }))
});
```

what does `t.maybe(t.struct({...` *means*? A natural answer would be:

- `name` is required
- `street` and `city` must be strings or.. both `null` (EDIT)

So can we infer the following rules?

> a maybe struct is `null` if and only if all its fields are `Nil`

> a maybe list is `null` if and only if all its elements are `Nil`

I'd love to hear your opinion on this

/cc @VinSpee

**Comment by volkanunsal** [comment]
> age and weight must be numbers or.. both null

I think you meant `street` and `city`? ;) In practice, I can see myself using this. So why not?

**Comment by gcanti** [comment]
opss...thanks @volkanunsal, fixed

**Comment by VinSpee** [comment]
I 100% agree. This is exactly what I would expect 



(null)

**Comment by gcanti** [comment]
About lists, perhaps it's better:

> a maybe list is null if and only if its length === 0

**Comment by gcanti** [comment]
Released in v0.6.10 and v0.7.6
